core/trino-main/src/test/java/io/trino/sql/planner/optimizations/TestAddExchangesPlans.java

```diff
@@ -978,6 +990,210 @@ public void testJoinNotExactlyPartitioned()
+                                .right(
+                                        exchange(LOCAL, GATHER, SINGLE_DISTRIBUTION,
+                                                exchange(REMOTE, REPLICATE, FIXED_BROADCAST_DISTRIBUTION,
+                                                        project(
+                                                                tableScan("region", ImmutableMap.of("regionkey", "regionkey")))))))));
+    }
+
+    @Test
+    public void testUnionAllAboveBroadcastJoin()
+    {
+        assertDistributedPlan(
+                """
+                    SELECT regionkey FROM nation UNION ALL (SELECT nationkey FROM nation n JOIN region r on r.regionkey = n.nationkey)
+                """,
+                noJoinReordering(),
+                anyTree(
+                        exchange(REMOTE, GATHER, SINGLE_DISTRIBUTION,
+                                tableScan("nation"),
+                                join(INNER, join -> join
+                                        .equiCriteria("nationkey", "regionkey")
+                                        .left(
+                                                project(
+                                                        node(FilterNode.class,
+                                                                tableScan("nation", ImmutableMap.of("nationkey", "nationkey")))))
+                                        .right(
+                                                exchange(LOCAL, GATHER, SINGLE_DISTRIBUTION,
+                                                        exchange(REMOTE, REPLICATE, FIXED_BROADCAST_DISTRIBUTION,
+                                                                project(
+                                                                        tableScan("region", ImmutableMap.of("regionkey", "regionkey"))))))))));
+    }
+
+    @Test
+    public void testAggregationAboveUnionAllCrossJoined()
+    {
+        assertDistributedPlan(
+                """
+                    SELECT sum(nationkey) FROM (SELECT nationkey FROM nation UNION ALL SELECT nationkey FROM nation), region
+                """,
+                noJoinReordering(),
+                anyTree(
+                        join(INNER, join -> join
+                                .left(
+                                        exchange(LOCAL, REPARTITION, FIXED_ARBITRARY_DISTRIBUTION,
+                                                tableScan("nation", ImmutableMap.of("nationkey", "nationkey")),
+                                                tableScan("nation")))
+                                .right(
+                                        exchange(LOCAL, GATHER, SINGLE_DISTRIBUTION,
+                                                exchange(REMOTE, REPLICATE, FIXED_BROADCAST_DISTRIBUTION,
+                                                        tableScan("region")))))));
+    }
+
+    @Test
+    public void testGroupedAggregationAboveUnionAllCrossJoined()
```

Comment:
Here, I have two tests: 
`testGroupedAggregationAboveUnionAll`, `testGroupedAggregationAboveUnionAllCrossJoined`. 
They are complementar because they expose case when parent is `SINGLE` / `DISTRIBUTED`.